core/lib/snapshots_applier/src/lib.rs

```diff
@@ -0,0 +1,356 @@
+pub(crate) enum InitialStage {
+    FetchMetadataFromMainNode,
+    ApplyFactoryDeps,
+}
+
+#[derive(Debug, Metrics)]
+#[metrics(prefix = "snapshots_creator")]
+pub(crate) struct SnapshotsApplierMetrics {
+    /// Number of chunks in the applied snapshot. Set when snapshots applier starts.
+    pub storage_logs_chunks_count: Gauge<usize>,
+
+    /// Number of chunks left to apply.
+    pub storage_logs_chunks_left_to_process: Gauge<usize>,
+
+    /// Total latency of applying snapshot.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub snapshot_applying_duration: Histogram<Duration>,
+
+    /// Latency of initial recovery operation split by stage.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub initial_stage_duration: Family<InitialStage, Histogram<Duration>>,
+
+    /// Latency of storage log chunk processing split by stage.
+    #[metrics(buckets = Buckets::LATENCIES, unit = Unit::Seconds)]
+    pub storage_logs_chunks_duration: Family<StorageLogsChunksStage, Histogram<Duration>>,
+}
+
+#[vise::register]
+pub(crate) static METRICS: vise::Global<SnapshotsApplierMetrics> = vise::Global::new();
+
+#[derive(thiserror::Error, Debug)]
+pub enum SnapshotsApplierError {
+    #[error("canceled")]
+    Canceled(String),
+    #[error(transparent)]
+    Fatal(#[from] anyhow::Error),
+    #[error("retryable")]
+    Retryable(String),
+}
+pub struct SnapshotsApplier<'a> {
+    connection_pool: &'a ConnectionPool,
+    blob_store: Box<dyn ObjectStore>,
+    applied_snapshot_status: SnapshotRecoveryStatus,
+}
+
+impl From<ObjectStoreError> for SnapshotsApplierError {
+    fn from(error: ObjectStoreError) -> Self {
+        match error {
+            ObjectStoreError::KeyNotFound(_) => SnapshotsApplierError::Fatal(error.into()),
+            ObjectStoreError::Serialization(_) => SnapshotsApplierError::Fatal(error.into()),
+            ObjectStoreError::Other(_) => SnapshotsApplierError::Retryable(
+                "An error occurred while accessing object store".to_string(),
+            ),
+        }
+    }
+}
+
+impl From<SqlxError> for SnapshotsApplierError {
+    fn from(err: SqlxError) -> Self {
+        match err {
+            SqlxError::Database(_)
+            | SqlxError::RowNotFound
+            | SqlxError::ColumnNotFound(_)
+            | SqlxError::Configuration(_)
+            | SqlxError::TypeNotFound { type_name: _ } => SnapshotsApplierError::Fatal(err.into()),
+            _ => SnapshotsApplierError::Retryable(format!(
+                "An error occured when accessing DB: {err}"
+            )),
+        }
+    }
+}
+
+impl From<RpcError> for SnapshotsApplierError {
+    fn from(err: RpcError) -> Self {
+        match err {
+            Error::Transport(_) | Error::RequestTimeout | Error::RestartNeeded(_) => {
+                SnapshotsApplierError::Retryable(
+                    "An error occurred while accessing main node".to_string(),
+                )
```

Comment:
I've refactored so that retryable error contains error instead of string. About first point, there are only 3 places in which I construct Canceled, so I don't think it's worth it.